README.md
# `vecbit` – Managing Memory Bit by Bit

[![Crate][crate_img]][crate]
[![Documentation][docs_img]][docs]
[![License][license_img]][license_file]
[![Continuous Integration][travis_img]][travis]
[![Code Coverage][codecov_img]][codecov]
[![Crate Downloads][downloads_img]][crate]
[![Crate Size][loc_img]][loc]

## Summary

`vecbit` enables refining memory manipulation from single-byte precision to
single-bit precision. The bit-precision pointers in this crate allow creation of
more powerful bit-masks, set arithmetic, and I/O packet processing.

The core export of this crate is the type `SliceBit`. This type is a region of
memory with individually-addressable bits. It is accessed by standard Rust
references: `&SliceBit` and `&mut SliceBit`. These references are able to
describe and operate on regions that start and end on any arbitrary bit address,
regardless of alignment to a byte or processor word.

Rust provides three types to manipulate a sequence of memory: `&[T]`/`&mut [T]`
to borrow a region, `Box<[T]>` to statically own a region, and `Vec<T>` to
dynamically own a region. `vecbit` provides parallel types for each: `&SliceBit`
and `&mut SliceBit` borrow, `BitBox` statically owns, and `VecBit` dynamically
owns. These types mirror the relationships and APIs, including inherent methods
and trait implementations, that are found in the standard library.

## What Makes `vecbit` Different Than All The Other Bit Vector Crates

The most significant differences are that `vecbit` provides arbitrary bit
ordering through the `Cursor` trait, and provides a full-featured slice type.
Other crates have fixed orderings, and are often unable to produce slices that
begin at any arbitrary bit.

Additionally, the `vecbit` types implement the full extent of the standard
library APIs possible, in both inherent methods and trait implementations.

The `vecbit` types’ handles are exactly the size of their standard library
counterparts, while the other crates carry bit index information in separate
fields, making their handles wider. Depending on your needs, this may sway your
opinion in either direction. `vecbit` is more compact, but mangles the internal
pointer representation and requires more complex logic to use the bit region,
while other crates’ types are larger, but have more straightforward internal
logic.

## Why Would You Use This

- You need to directly control a bitstream’s representation in memory.
- You need to do unpleasant things with I/O communications protocols.
- You need a list of `bool`s that doesn’t waste 7 bits for every bit used.
- You need to do set arithmetic, or numeric arithmetic, on those lists.
- You are running a find/replace command on your repository from `&[bool]` to
  `&SliceBit`, or `Vec<bool>` to `VecBit`, and expect minimal damage as a
  result.

## Why *Wouldn’t* You Use This

- Your concern with the memory representation of bitsets includes sequence
  compression. `SliceBit` performs absolutely no compression, and maps bits
  directly into memory. Compressed bit sets can be found in other crates, such
  as the [`compacts`] crate, which uses the [Roaring BitSet] format.
- You want discontiguous data structures, such as a hash table, a tree, or any
  other hallmark of computer science beyond the flat array. `vecbit` does not,
  and will not, accomplish this. You may be able to use `vecbit`’s flat
  structures beneath a type wrapper which handles index processing, but `vecbit`
  types are incapable of accomplishing this task themselves. Also, I don’t know
  how any of those data structures work.

## Usage

**Minimum Rust Version**: `1.36.0`

The `1.36` release of Rust stabilized the `alloc` crate, allowing allocating
features (such as the `VecBit` type) to be used in `#![no_std]` environments
with the stable compiler series.

### Symbol Import

```toml
# Cargo.toml
[dependencies]
vecbit = "0.15"
```

`vecbit` is highly modular, and requires several items to function correctly.
The simplest way to use it is via prelude glob import:

```rust
use vecbit::prelude::*;
```

This imports the following symbols:

- `BigEndian`
- `BitBox` (only when an allocator is present)
- `SliceBit`
- `BitStore`
- `VecBit` (only when an allocator is present)
- `Bits`
- `BitsMut`
- `Cursor`
- `LittleEndian`
- `bitbox!` (only when an allocator is present)
- `vecbit!` (only when an allocator is present)

If you do not want these names imported directly into the local scope –
`Cursor`, `BigEndian`, and `LittleEndian` are likely culprits for name collision
– then you can import the prelude with a scope guard:

```rust
use vecbit::prelude as bv;
```

and those symbols will all be available only with a `bv::` prefix.

### Cargo Features

`vecbit` uses Cargo features to conditionally control some behavior.

The most prominent such behavior is one that cannot be controlled by Cargo
configuration: `u64` is only usable with this library when targeting a 64-bit
system. 32-bit system targets are only permitted to use `u8`, `u16`, and `u32`.

#### Atomic Behavior

`vecbit` uses atomic read/modify/write instructions by default. This is
necessary to avoid data races in `&mut SliceBit` operations without using
heavier synchronization mechanisms. If your target does not support Rust’s
`AtomicU*` types, or you do not want to use atomic RMW instructions, you may
disable the `atomic` feature:

```toml
# Cargo.toml

[dependencies.vecbit]
default-features = false
features = [
  # "atomic",
  "std",
]
```

#### Allocator Support

The two owning structures, `BitBox` and `VecBit`, require the presence of an
allocator. As `vecbit` is written specifically for use cases where an allocator
may not exist, this dependence can be disabled. `vecbit` is
`#![no_std]`-compatible once the `std` feature is disabled. It is not a design
goal to be `#![no_core]`-compatible.

```toml
# Cargo.toml

[dependencies.vecbit]
default-features = false
features = [
  "atomic",
  # "std",
]
```

If you are working in a `#![no_std]` environment that does have an allocator
available, you can reënable allocator support with the `alloc` feature:

```toml
# Cargo.toml

[dependencies.vecbit]
default-features = false # disables "std"
features = ["alloc"] # enables the allocator
```

This uses `#![feature(alloc)]`, which requires the nightly compiler.

#### Serde Support

De/serialization of bit slices is implemented through the `serde` crate. This
functionality is governed by both the `serde` feature and the `std` feature.

By default, when `serde` is enabled, `SliceBit`, `BitBox`, and `VecBit` all gain
the `Serialize` trait, and `BitBox` and `VecBit` gain the `Deserialize` trait.

When `std` is disabled, the `BitBox` and `VecBit` types are removed, leaving
only `SliceBit` with `Serialize`.

```toml
# Cargo.toml

[dependencies.vecbit]
features = ["serde"]
```

### Data Structures

`vecbit`’s three data structures are `&SliceBit`, `BitBox`, and `VecBit`. Each
of these types takes two type parameters, which I have elided previously.

The first type parameter is the `Cursor` trait. This trait governs how a bit
index maps to a bit position in the underlying memory. This parameter defaults
to the `BigEndian` type, which counts from the most significant bit first to the
least significant bit last. The `LittleEndian` type counts in the opposite
direction.

The second type parameter is the `BitStore` trait. This trait abstracts over the
Rust fundamental types `u8`, `u16`, and `u32`. On 64-bit targets, `u64` is also
available. This parameter defaults to `u8`, which acts on individual bytes.

These traits are both explained in the next section.

`&SliceBit<C: Cursor, T: BitStore>` is an immutable region of memory,
addressable at bit precision. This has all the inherent methods of Rust’s slice
primitive, `&[bool]`, and all the trait implementations. It has additional
methods which are specialized to its status as a slice of individual bits.

`&mut SliceBit<C: Cursor, T: BitStore>` is a mutable region of memory. This
functions identically to `&mut [bool]`, with the exception that `IndexMut` is
impossible: you cannot write `bitslice[index] = bit;`. This restriction is
sidestepped with the C++-style method `at`: `*bitslice.at(index) = bit;` is the
shim for write indexing.

The slice references have no restrictions on the alignment of their start or
end bits.

The owning references, described below, will always begin their slice aligned to
the edge of their `T: BitStore` type parameter. While this is not strictly
required by the implementation, it is convenient for ensuring that the
allocation pointer is preserved.

`BitBox<C: Cursor, T: BitStore>` is a `&mut SliceBit<C: Cursor, T: BitStore>` in
owned memory. It has few useful methods and no trait implementations of its own.
It is only capable of taking a bit slice into owned memory.

`VecBit<C: Cursor, T: BitStore>` is a `BitBox` that can adjust its allocation
size. It follows the inherent and trait API of the standard library’s `Vec`
type.

The API for these types is deliberately uninteresting. They are written to be as
close to drop-in replacements for the standard library types as possible. The
end goal of `vecbit` is that you should be able to adopt it by running three
`sed` find/replace commands on your repository. This is not literally possible,
but the work required for replacement is intended to be minimal.

### Traits

`vecbit` generalizes its behavior through the use of traits. In order to
optimize performance, these traits are *not* object-safe, and may *not* be used
as `dyn Trait` patterns for type erasure. Refactoring `vecbit` to support type
erasure would require significantly rewriting core infrastructure, and this is
not a design goal. I am willing to consider it if demand is shown, but I am not
going to proactively pursue it.

#### `Cursor`

The `Cursor` trait is an open-ended trait, that you are free to implement
yourself. It has one required function: `fn at<T: BitStore>(BitIdx) -> BitPos`.

This function translates a semantic index to an electrical position. `vecbit`
provides two implementations for you: `BigEndian` and `LittleEndian`, described
above. The invariants this function must uphold are listed in its documentation.

#### `BitStore`

The `BitStore` trait is sealed, and may only be implemented by this library. It
is used to abstract over the Rust fundamentals `u8`, `u16`, `u32`, and (on
64-bit systems) `u64`.

Your choice in fundamental types governs how the `Cursor` type translates
indices, and how the memory underneath your slice is written. The document
`doc/Bit Patterns.md` enumerates the effect of the `Cursor` and `BitStore`
combinations on raw memory.

If you are using `vecbit` to perform set arithmetic, and you expect that your
sets will have more full elements in the interior than partially-used elements
on the front and back edge, it is advantageous to use the local CPU word. The
`SliceBit` operations which traverse the slice are required to perform
bit-by-bit crawls on partially-used elements, but are able to use whole-word
instructions on full elements. The latter is a marked acceleration. The local
CPU word size (accessible through the type alias `vecbit::store::Local`) is the
default storage type for all data structures in the library.

If you are using `vecbit` to perform I/O packet manipulation, you should use the
fundamental and ordering that match your protocol specification.

#### `Bits` and `BitsMut`

The `Bits` and `BitsMut` traits are entry interfaces to the `SliceBit` types.
These are equivalent to the `AsRef` and `AsMut` reference conversion traits in
the standard library, and should be used as such.

These traits are implemented on the Rust fundamentals that implement `BitStore`
(`uN`), on slices of those fundamentals (`[uN]`), and the first thirty-two
arrays of them (`[uN; 0]` to `[uN; 32]`). Each implementation of these traits
causes a linear expansion of compile time, and going beyond thirty-two both
surpasses the standard library’s manual implementation limits, and is a
denial-of-service attack on each rebuild.

These traits are left open so that if you need to implement them on wider
arrays, you are able to do so.

You can use these traits to attach `.bits::<C: Cursor>()` and
`.bits_mut::<C: Cursor>()` conversion methods to any implementor, and gain
access to a `SliceBit` over that type, or to bound a generic function similar to
how the standard library uses `AsRef<Path>`:

```rust
let mut base = [0u8; 8];
let bits = base.bits_mut::<LittleEndian>();
//  bits is now an `&mut SliceBit<LittleEndian, u8>`
println!("{}", bits.len()); // 64

fn operate_on_bits(mut data: impl BitsMut) {
  let bits = data.bits_mut::<BigEndian>();
  //  `bits` is now an `&mut SliceBit<BigEndian, _>`
}
```

### Macros

The `bitbox!` and `vecbit!` macros allow convenient production of their
eponymous types, equivalent to the `vec!` macro in the standard library.

These macros accept an optional cursor token, an optional type token, and either
a list of bits or a single bit and a repetition counter.

Because these are standard macros, not proc-macros, they do not yet produce
well-optimized expanded code.

These macros are more thoroughly explained, including a list of all available
use syntaxes, in their documentation.

## Example Usage

This snippet runs through a selection of library functionality to demonstrate
behavior. It is deliberately not representative of likely usage.

```rust
extern crate vecbit;

use vecbit::prelude::*;

use std::iter::repeat;

fn main() {
    let mut bv = vecbit![BigEndian, u8; 0, 1, 0, 1];
    bv.reserve(8);
    bv.extend(repeat(false).take(4).chain(repeat(true).take(4)));

    //  Memory access
    assert_eq!(bv.as_slice(), &[0b0101_0000, 0b1111_0000]);
    //                   index 0 -^               ^- index 11
    assert_eq!(bv.len(), 12);
    assert!(bv.capacity() >= 16);

    //  Stack operations
    bv.push(true);
    bv.push(false);
    bv.push(true);

    assert!(bv[12]);
    assert!(!bv[13]);
    assert!(bv[14]);
    assert!(bv.get(15).is_none());

    bv.pop();
    bv.pop();
    bv.pop();

    //  Set operations. These deliberately have no effect.
    bv &= repeat(true);
    bv = bv | repeat(false);
    bv ^= repeat(true);
    bv = !bv;

    //  Arithmetic operations
    let one = vecbit![BigEndian, u8; 1];
    bv += one.clone();
    assert_eq!(bv.as_slice(), &[0b0101_0001, 0b0000_0000]);
    bv -= one.clone();
    assert_eq!(bv.as_slice(), &[0b0101_0000, 0b1111_0000]);

    //  Borrowing iteration
    let mut iter = bv.iter();
    //  index 0
    assert_eq!(iter.next().unwrap(), false);
    //  index 11
    assert_eq!(iter.next_back().unwrap(), true);
    assert_eq!(iter.len(), 10);
}
```

Immutable and mutable access to the underlying memory is provided by the `AsRef`
and `AsMut` implementations, so the `VecBit` can be readily passed to transport
functions.

`VecBit` implements `Borrow` down to `SliceBit`, and `SliceBit` implements
`ToOwned` up to `VecBit`, so they can be used in a `Cow` or wherever this API
is desired. Any case where a `Vec`/`[T]` pair cannot be replaced with a
`VecBit`/`SliceBit` pair is a bug in this library, and a bug report is
appropriate.

`VecBit` can relinquish its owned memory with `.into_vec()` or
`.into_boxed_slice()`, and `SliceBit` can relinquish its borrow by going out
of scope.

## Warnings

The `SliceBit` type is able to cause memory aliasing, as multiple independent
`&mut SliceBit` instances may use the same underlying memory. This crate takes
care to ensure that all observed behavior is exactly as expected, without any
side effects.

The `SliceBit` methods only use whole-element instructions when the slice spans
the full width of the element; when the slice has only partial use, the methods
crawl each bit individually. This is slower on most architectures, but
guarantees safety.

Race conditions are avoided through use of the atomic read/modify/write
instructions stabilized in `1.34.0`, as described above.

## Planned Features

Contributions of items in this list are *absolutely* welcome! Contributions of
other features are also welcome, but I’ll have to be sold on them.

- Creation of specialized pointers `Rc<SliceBit>` and `Arc<SliceBit>`.
- Procedural macros for `vecbit!` and `bitbox!`
- An FFI module, and bindings from other languages.

[codecov]: https://codecov.io/gh/sunjay/vecbit "Code Coverage"
[codecov_img]: https://img.shields.io/codecov/c/github/sunjay/vecbit.svg?logo=codecov "Code Coverage Display"
[crate]: https://crates.io/crates/vecbit "Crate Link"
[crate_img]: https://img.shields.io/crates/v/vecbit.svg?logo=rust "Crate Page"
[docs]: https://docs.rs/vecbit "Documentation"
[docs_img]: https://docs.rs/vecbit/badge.svg "Documentation Display"
[downloads_img]: https://img.shields.io/crates/dv/vecbit.svg?logo=rust "Crate Downloads"
[license_file]: https://github.com/sunjay/vecbit/blob/master/LICENSE.txt "License File"
[license_img]: https://img.shields.io/crates/l/vecbit.svg "License Display"
[loc]: https://github.com/sunjay/vecbit "Repository"
[loc_img]: https://tokei.rs/b1/github/sunjay/vecbit "Repository Size"
[travis]: https://travis-ci.org/sunjay/vecbit "Travis CI"
[travis_img]: https://img.shields.io/travis/sunjay/vecbit.svg?logo=travis "Travis CI Display"
[`compacts`]: https://crates.io/crates/compacts
[Roaring BitSet]: https://arxiv.org/pdf/1603.06549.pdf


CHANGELOG.md
# Changelog

All notable changes will be documented in this file.

This document is written according to the [Keep a Changelog][kac] style.

## 0.16.0

### Added

- `Cursor` now provides a `mask` function, which produces a one-hot mask usable
  for direct memory access. Implementors of `Cursor` may use the default, or
  provide their own.
- `Bits` and `BitsMut` renamed their methods to `bits` and `bits_mut`,
  respectively; `as_bitslice` and `as_mut_bitslice` are marked deprecated and
  will be removed in `0.17`.
- The `BitField` trait allows `SliceBit<BigEndian, _>` and
  `SliceBit<LittleEndian, _>` to provide behavior analagous to bitfields in C
  and C++ `struct` definitions. This trait provides `load` and `store` methods
  on `SliceBit`s with those two `Cursor`s which allow for parallel access to the
  underlying memory. This trait is currently not able to be implemented by
  downstream crates; this restriction may ease in the future.
- The `cursor::Local` type alias is a default bit ordering. Big-endian targets
  set it to `cursor::BigEndian`; all other targets set it to
  `cursor::LittleEndian`.
- The `store::Word` type alias is a default unit size. Targets with 32-bit CPU
  words set it to `u32`; 64-bit CPU word targets set it to `u64`; all other
  targets set it to `u8`.

### Changed

- The default order and storage type parameters for all type constructors in the
  library have been changed. This means that `SliceBit`, `BitBox`, `VecBit`, and
  the `bitbox!` and `vecbit!` macros, are all changing the produced type if you
  have not specified their ordering and storage. The new default storage type is
  the target CPU word (`u32` on 32-bit systems, `u64` on 64-bit, `u8` on other)
  and the new default order type is the target byte ordering (`BigEndian` on
  big-endian, `LittleEndian` on little-endian and unknown).

  This change is expected to break dependent crates. The fix is straightforward:
  specify the types produced by this crate’s constructors, or adapt the types
  that receive them.

  This change was made in order to provide performance advantages by using the
  native CPU word size, and to ease choice of a bit ordering in usages that do
  not particularly care about the underlying memory’s appearance.
- The internal process that translates `SliceBit` operations into access
  operations on underlying memory has been rewritten. Production of contended
  references to bare fundamentals is now forbidden, and all access is mediated
  through either atomic (default) or `Cell` types.
- Bit indexing is more firmly encoded in the type system, which reduces the load
  of runtime assertions.
- `SliceBit::as_slice` excludes partial edge elements. `BitBox` and `VecBit` do
  not.

### Removed

- `SliceBit::change_cursor` and `change_cursor_mut` allow incorrectly aliasing
  memory with different slice handles, because there is no way for them to
  compute the electrical positions they govern and then construct new indices in
  the target `Cursor` that match those positions.

  Example:

  ```rust
  use vecbit::prelude::*;
  let mut elt = 0u8;
  let bits = elt.bits_mut::<BigEndian>();
  let (head, tail) = bits.split_at_mut(4);
  let tail = tail.change_cursor_mut::<LittleEndian>();
  ```

  `head` now points at the first four bits in big-endian order, and tail at the
  last four bits in little-endian order, and these indices all map to the high
  nibble of `elt`. `head` and `tail` mutably alias.

  These functions are retained in `BitBox` and `VecBit`, as those types do not
  allow memory contention.

## 0.15.2

### Changed

The `vecbit![bit; rep]` construction macro has its implementation rewritten to
be much faster. This fault was reported by GitHub user [@caelunshun] in
[Issue #28].

## 0.15.1

### Removed

The `Send` implementation on `SliceBit` is removed when the `atomic` feature is
disabled.

While the `x86` architecture provides hardware guarantees that a
read/modify/write instruction sequence will update all other views of the
referent data, this is a property of the specific underlying machine and *not*
a property of the Rust abstract machine as interpreted by the compiler and LLVM.
As such, while `&mut SliceBit` references that alias the same underlying memory
element will not collide with each other in practice, they still must use atomic
operations in order to satisfy the Rust abstract machine.

The atomic feature is provided by default, and users must explicitly disable it
in order to disable atomic instruction access and thus remove the `Send` impl
allowing `&mut SliceBit` to cross threads.

Because this change does not affect the default interface exported by the crate,
I have decided to make this a patch release rather than bump the minor version.

## 0.15.0

### Changed

The minimum compiler version was increased to `1.36.0`, which stabilized the
`alloc` crate. As such, the `#![feature(alloc)]` flag has been removed from the
library, and usage as `--no-default-features --features alloc` may safely use
allocation on the stable compiler series.

As `alloc` is available on a stable compiler, the `alloc` *crate feature* has
been made a strict dependency of the `std` crate feature.

Use of `--no-default-features` continues to set the crate in `#![no_std]` mode,
with no allocation support. `--no-default-features --features alloc` adds a
dependency on `alloc`, and the allocating types. The `std` feature alone now
*only* adds operating-system interfaces, such as `io::Write`.

`std` depends on `alloc`, so using the `std` feature will pull in allocation.

## 0.14.0

### Added

- `add_assign_reverse` on `SliceBit` and `VecBit`, and `add_reverse` on
  `BitBox` and `VecBit`.

  These methods perform left-to-right addition, propagating the carry from the
  0th bit in the sequence to the nth. On `SliceBit`, `add_assign_reverse`
  returns the carry-out bit. On `VecBit`, `add_assign_reverse` and `add_reverse`
  push the carry-out to the right end of the vector.


src/serdes.rs
/*! `serde`-powered de/serialization

This module implements the Serde traits for the `vecbit` types, as possible.

Without an allocator, only `SliceBit` exists, and can only implement
`Serialize`. With an allocator, the `BitBox` and `VecBit` types exist, and are
able to implement `Deserialize` as well.
!*/

#![cfg(all(feature = "serde"))]

use crate::{
	cursor::Cursor,
	slice::SliceBit,
	store::BitStore,
};

#[cfg(feature = "alloc")]
use crate::{
	boxed::BitBox,
	pointer::BitPtr,
	vec::VecBit,
};

#[cfg(feature = "alloc")]
use core::{
	cmp,
	convert::TryInto,
	fmt::{
		self,
		Formatter,
	},
	marker::PhantomData,
	mem,
};

use serde::{
	Serialize,
	ser::{
		Serializer,
		SerializeStruct,
	},
};

#[cfg(feature = "alloc")]
use serde::{
	Deserialize,
	de::{
		self,
		Deserializer,
		Error,
		MapAccess,
		SeqAccess,
		Unexpected,
		Visitor,
	},
};

/// A Serde visitor to pull `BitBox` data out of a serialized stream
#[cfg(feature = "alloc")]
#[derive(Clone, Copy, Default, Debug)]
pub struct BitBoxVisitor<'de, C, T>
where C: Cursor, T: BitStore + Deserialize<'de> {
	_cursor: PhantomData<C>,
	_storage: PhantomData<&'de T>,
}

#[cfg(feature = "alloc")]
impl<'de, C, T> BitBoxVisitor<'de, C, T>
where C: Cursor, T: BitStore + Deserialize<'de> {
	fn new() -> Self {
		BitBoxVisitor { _cursor: PhantomData, _storage: PhantomData }
	}
}

#[cfg(feature = "alloc")]
impl<'de, C, T> Visitor<'de> for BitBoxVisitor<'de, C, T>
where C: Cursor, T: BitStore + Deserialize<'de> {
	type Value = BitBox<C, T>;

	fn expecting(&self, fmt: &mut Formatter) -> fmt::Result {
		fmt.write_str("A BitSet data series")
	}

	/// Visit a sequence of anonymous data elements. These must be in the order
	/// `usize', `u8`, `u8`, `[T]`.
	fn visit_seq<V>(self, mut seq: V) -> Result<Self::Value, V::Error>
	where V: SeqAccess<'de> {
		let head: u8 = seq.next_element()?
			.ok_or_else(|| de::Error::invalid_length(0, &self))?;
		let bits: usize = seq.next_element()?
			.ok_or_else(|| de::Error::invalid_length(1, &self))?;
		let data: Box<[T]> = seq.next_element()?
			.ok_or_else(|| de::Error::invalid_length(2, &self))?;

		let bitptr = BitPtr::new(
			data.as_ptr(),
			head.try_into().map_err(|_| Error::invalid_value(
				Unexpected::Unsigned(head as u64),
				&self,
			))?,
			bits,
		);
		mem::forget(data);
		Ok(unsafe { BitBox::from_raw(bitptr) })
	}

	/// Visit a map of named data elements. These may be in any order, and must
	/// be the pairs `head: u8`, `bits: usize`, and `data: [T]`.
	fn visit_map<V>(self, mut map: V) -> Result<Self::Value, V::Error>
	where V: MapAccess<'de> {
		let mut head: Option<u8> = None;
		let mut bits: Option<usize> = None;
		let mut data: Option<Box<[T]>> = None;

		while let Some(key) = map.next_key()? {
			match key {
				"head" => if head.replace(map.next_value()?).is_some() {
					return Err(de::Error::duplicate_field("head"));
				},
				"bits" => if bits.replace(map.next_value()?).is_some() {
					return Err(de::Error::duplicate_field("bits"));
				},
				"data" => if data.replace(map.next_value()?).is_some() {
					return Err(de::Error::duplicate_field("data"));
				},
				f => return Err(de::Error::unknown_field(
					f, &["head", "bits", "data"]
				)),
			}
		}
		let head = head.ok_or_else(|| de::Error::missing_field("head"))?;
		let bits = bits.ok_or_else(|| de::Error::missing_field("bits"))?;
		let data = data.ok_or_else(|| de::Error::missing_field("data"))?;

		let bitptr = BitPtr::new(
			data.as_ptr(),
			head.try_into().map_err(|_| Error::invalid_value(
				Unexpected::Unsigned(head as u64),
				&self,
			))?,
			cmp::min(bits, data.len() * T::BITS as usize),
		);
		mem::forget(data);
		Ok(unsafe { BitBox::from_raw(bitptr) })
	}
}

#[cfg(feature = "alloc")]
impl<'de, C, T> Deserialize<'de> for BitBox<C, T>
where C: Cursor, T: 'de + BitStore + Deserialize<'de> {
	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
	where D: Deserializer<'de> {
		deserializer
			.deserialize_struct(
				"BitSet",
				&["head", "bits", "data"],
				BitBoxVisitor::new(),
			)
	}
}

#[cfg(feature = "alloc")]
impl<'de, C, T> Deserialize<'de> for VecBit<C, T>
where C: Cursor, T: 'de + BitStore + Deserialize<'de> {
	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
	where D: Deserializer<'de> {
		BitBox::deserialize(deserializer).map(Into::into)
	}
}

impl<C, T> Serialize for SliceBit<C, T>
where C: Cursor, T: BitStore + Serialize, T::Access: Serialize {
	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
	where S: Serializer {
		let head = self.bitptr().head();
		let mut state = serializer.serialize_struct("BitSet", 3)?;

		state.serialize_field("head", &*head)?;
		state.serialize_field("bits", &(self.len() as u64))?;
		state.serialize_field("data", self.as_total_slice())?;

		state.end()
	}
}

#[cfg(feature = "alloc")]
impl<C, T> Serialize for BitBox<C, T>
where C: Cursor, T: BitStore + Serialize, T::Access: Serialize {
	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
	where S: Serializer {
		SliceBit::serialize(&*self, serializer)
	}
}

#[cfg(feature = "alloc")]
impl<C, T> Serialize for VecBit<C, T>
where C: Cursor, T: BitStore + Serialize, T::Access: Serialize {
	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
	where S: Serializer {
		SliceBit::serialize(&*self, serializer)
	}
}

#[cfg(test)]
mod tests {
	use crate::prelude::*;
	use serde_test::{
		Token,
		assert_ser_tokens,
	};
	#[cfg(feature = "alloc")]
	use serde_test::assert_de_tokens;

	macro_rules! bvtok {
		( s $elts:expr, $head:expr, $bits:expr, $ty:ident $( , $data:expr )* ) => {
			&[
				Token::Struct { name: "BitSet", len: 3, },
				Token::Str("head"), Token::U8( $head ),
				Token::Str("bits"), Token::U64( $bits ),
				Token::Str("data"), Token::Seq { len: Some( $elts ) },
				$( Token:: $ty ( $data ), )*
				Token::SeqEnd,
				Token::StructEnd,
			]
		};
		( d $elts:expr, $head:expr, $bits:expr, $ty:ident $( , $data:expr )* ) => {
			&[
				Token::Struct { name: "BitSet", len: 3, },
				Token::BorrowedStr("head"), Token::U8( $head ),
				Token::BorrowedStr("bits"), Token::U64( $bits ),
				Token::BorrowedStr("data"), Token::Seq { len: Some( $elts ) },
				$( Token:: $ty ( $data ), )*
				Token::SeqEnd,
				Token::StructEnd,
			]
		};
	}

	#[test]
	fn empty() {
		let slice = SliceBit::<BigEndian, u8>::empty();

		assert_ser_tokens(&slice, bvtok![s 0, 0, 0, U8]);

		#[cfg(feature = "alloc")]
		assert_de_tokens(&vecbit![], bvtok![ d 0, 0, 0, U8 ]);
	}

	#[cfg(feature = "alloc")]
	#[test]
	fn small() {
		let bv = vecbit![BigEndian, u8; 1; 5];
		let bs = &bv[1 ..];
		assert_ser_tokens(&bs, bvtok![s 1, 1, 4, U8, 0b1111_1000]);

		let bv = vecbit![LittleEndian, u16; 1; 12];
		assert_ser_tokens(&bv, bvtok![s 1, 0, 12, U16, 0b00001111_11111111]);

		let bb: BitBox<_, _> = vecbit![LittleEndian, u32; 1; 10].into();
		assert_ser_tokens(&bb, bvtok![s 1, 0, 10, U32, 0x00_00_03_FF]);
	}

	#[cfg(feature = "alloc")]
	#[test]
	fn wide() {
		let src: &[u8] = &[0, !0];
		let bs = src.bits::<Local>();
		assert_ser_tokens(&(&bs[1 .. 15]), bvtok![s 2, 1, 14, U8, 0, !0]);
	}

	#[cfg(feature = "alloc")]
	#[test]
	fn deser() {
		let bv = vecbit![BigEndian, u8; 0, 1, 1, 0, 1, 0];
		assert_de_tokens(&bv, bvtok![d 1, 0, 6, U8, 0b0110_1000]);
		//  test that the bits outside the bits domain don't matter in deser
		assert_de_tokens(&bv, bvtok![d 1, 0, 6, U8, 0b0110_1001]);
		assert_de_tokens(&bv, bvtok![d 1, 0, 6, U8, 0b0110_1010]);
		assert_de_tokens(&bv, bvtok![d 1, 0, 6, U8, 0b0110_1011]);
	}
}
